core/dataset.cc

```diff
@@ -1138,7 +1138,7 @@ generateRowsWhere(const SqlBindingScope & scope,
     // Optimize for rowName() IN (constant, constant, constant)
     // Optimize for rowName() IN ROWS / IN KEYS (...)
     auto inExpression = dynamic_cast<const InExpression *>(&where);
-    if (inExpression) 
+    if (inExpression && !inExpression->isnegative) 
```

Comment:
This disables the optimization for `NOT IN`, isn't?  Shouldn't we handle it inside this optimized path instead?